File: governance/governance/reporters/github.py
"""GitHub / dashboard reporters for pipeline results."""

from __future__ import annotations

import os
from typing import Any

import httpx

from governance.models import Finding, PipelineReport


def format_markdown(report: PipelineReport) -> str:
    lines = [
        "## AI Code Guardrail Report",
        "",
        f"**Overall:** {'✅ PASSED' if report.passed else '❌ FAILED'}",
        f"**Blocking findings:** {report.summary.get('blocking_findings', 0)}",
        "",
    ]
    for step in report.steps:
        icon = "⏭" if step.skipped else ("✅" if step.passed else "❌")
        lines.append(f"### {icon} {step.name}")
        if step.skipped:
            lines.append(f"_Skipped:_ {step.skip_reason}")
            lines.append("")
            continue
        if not step.findings:
            lines.append("_No findings._")
            lines.append("")
            continue
        for f in step.findings:
            loc = ""
            if f.file:
                loc = f"`{f.file}`"
                if f.line:
                    loc += f":{f.line}"
                loc += " — "
            lines.append(f"- **{f.severity.value.upper()}** {loc}{f.message}")
            if f.suggestion:
                lines.append(f"  - Suggestion: {f.suggestion}")
        lines.append("")
    return "\n".join(lines)


def post_github_pr_comment(report: PipelineReport) -> dict[str, Any] | None:
    """Leave a PR comment summarizing the suite (uses GITHUB_TOKEN)."""
    token = os.getenv("GITHUB_TOKEN")
    repo = report.repo or os.getenv("GITHUB_REPOSITORY")
    pr_number = report.pr_number or _env_int("PR_NUMBER")
    if not token or not repo or not pr_number:
        return None

    body = format_markdown(report)
    url = f"https://api.github.com/repos/{repo}/issues/{pr_number}/comments"
    with httpx.Client(timeout=30.0) as client:
        resp = client.post(
            url,
            headers={
                "Authorization": f"Bearer {token}",
                "Accept": "application/vnd.github+json",
            },
            json={"body": body},
        )
        resp.raise_for_status()
        return resp.json()


def post_inline_comments(report: PipelineReport, commit_sha: str | None = None) -> int:
    """Post inline review comments for findings that have file+line."""
    token = os.getenv("GITHUB_TOKEN")
    repo = report.repo or os.getenv("GITHUB_REPOSITORY")
    pr_number = report.pr_number or _env_int("PR_NUMBER")
    sha = commit_sha or report.commit_sha or os.getenv("GITHUB_SHA")
    if not token or not repo or not pr_number or not sha:
        return 0

    posted = 0
    with httpx.Client(timeout=30.0) as client:
        for step in report.steps:
            for finding in step.findings:
                if not finding.file or not finding.line:
                    continue
                if finding.severity.value not in {"error", "critical"}:
                    continue
                # Prefer relative path from repo root
                path = finding.file
                if path.startswith("/"):
                    # best-effort strip to repo-relative
                    marker = f"/{repo.split('/')[-1]}/"
                    if marker in path:
                        path = path.split(marker, 1)[1]
                    elif "/workspace/" in path:
                        path = path.split("/workspace/", 1)[1]
                payload = {
                    "body": _inline_body(finding),
                    "commit_id": sha,
                    "path": path,
                    "line": finding.line,
                    "side": "RIGHT",
                }
                url = f"https://api.github.com/repos/{repo}/pulls/{pr_number}/comments"
                resp = client.post(
                    url,
                    headers={
                        "Authorization": f"Bearer {token}",
                        "Accept": "application/vnd.github+json",
                    },
                    json=payload,
                )
                if resp.status_code < 300:
                    posted += 1
    return posted


def post_to_dashboard(report: PipelineReport) -> dict[str, Any] | None:
    """POST pipeline results to the Step 7 review dashboard API.

    Soft-fails: dashboard outages must not fail an otherwise-green CI job.
    """
    endpoint = os.getenv("GOVERNANCE_DASHBOARD_URL")
    if not endpoint:
        return None
    secret = os.getenv("GOVERNANCE_DASHBOARD_SECRET", "")
    url = f"{endpoint.rstrip('/')}/api/reviews"
    try:
        with httpx.Client(timeout=30.0) as client:
            resp = client.post(
                url,
                headers={
                    "Content-Type": "application/json",
                    "X-Governance-Secret": secret,
                },
                json=report.model_dump(),
            )
            if resp.status_code == 401:
                return {
                    "ok": False,
                    "error": (
                        "401 Unauthorized — GOVERNANCE_DASHBOARD_SECRET in GitHub "
                        "Actions must exactly match GOVERNANCE_DASHBOARD_SECRET on "
                        f"the dashboard host ({endpoint.rstrip('/')})."
                    ),
                }
            resp.raise_for_status()
            return resp.json()
    except Exception as exc:  # noqa: BLE001 — never fail the suite on dashboard I/O
        return {"ok": False, "error": str(exc)}


QUIZ_STATUS_CONTEXT = "Governance Quiz"


def post_quiz_commit_status(
    report: PipelineReport,
    *,
    state: str = "pending",
    description: str = "Take the Step 6 comprehension quiz on the governance dashboard.",
) -> dict[str, Any] | None:
    """Create/replace the branch-protection commit status for the quiz gate.

    Soft-fails when token/repo/sha are missing. Dashboard later sets success
    after the human passes the quiz for this SHA.
    """
    token = os.getenv("GITHUB_TOKEN")
    repo = report.repo or os.getenv("GITHUB_REPOSITORY")
    sha = report.commit_sha or os.getenv("GITHUB_SHA")
    if not token or not repo or not sha:
        return None
    if state not in {"pending", "success", "failure", "error"}:
        state = "pending"

    dashboard = (os.getenv("GOVERNANCE_DASHBOARD_URL") or "").rstrip("/")
    payload: dict[str, Any] = {
        "state": state,
        "context": QUIZ_STATUS_CONTEXT,
        "description": description[:140],
    }
    if dashboard:
        payload["target_url"] = dashboard[:1024]

    url = f"https://api.github.com/repos/{repo}/statuses/{sha}"
    try:
        with httpx.Client(timeout=20.0) as client:
            resp = client.post(
                url,
                headers={
                    "Authorization": f"Bearer {token}",
                    "Accept": "application/vnd.github+json",
                    "X-GitHub-Api-Version": "2022-11-28",
                },
                json=payload,
            )
            if resp.status_code >= 300:
                return {
                    "ok": False,
                    "error": f"{resp.status_code} {resp.text[:200]}",
                }
            return resp.json()
    except Exception as exc:  # noqa: BLE001
        return {"ok": False, "error": str(exc)}


def _inline_body(finding: Finding) -> str:
    parts = [
        f"**[{finding.rule_id or finding.step}] {finding.severity.value.upper()}**",
        finding.message,
    ]
    if finding.suggestion:
        parts.append(f"\n💡 {finding.suggestion}")
    return "\n".join(parts)


def _env_int(name: str) -> int | None:
    raw = os.getenv(name)
    if not raw:
        return None
    try:
        return int(raw)
    except ValueError:
        return None

File: dashboard/src/lib/github.ts
/**
 * Safe GitHub URL helpers + Governance Quiz commit status.
 *
 * Branch protection can require context {@link QUIZ_STATUS_CONTEXT}.
 * CI sets it to pending; the dashboard sets success/failure after the quiz.
 */

const REPO_RE = /^([A-Za-z0-9_.-]+)\/([A-Za-z0-9_.-]+)$/;

/** Must match the name you require under branch protection. */
export const QUIZ_STATUS_CONTEXT = "Governance Quiz";

export type GithubRepoRef = {
  owner: string;
  name: string;
  full: string;
};

export type QuizStatusState = "pending" | "success" | "failure" | "error";

export function parseGithubRepo(repo: string | null | undefined): GithubRepoRef | null {
  if (!repo || typeof repo !== "string") return null;
  const match = REPO_RE.exec(repo.trim());
  if (!match) return null;
  return { owner: match[1], name: match[2], full: `${match[1]}/${match[2]}` };
}

export function parsePositiveInt(value: unknown): number | null {
  const n = typeof value === "number" ? value : Number(value);
  if (!Number.isInteger(n) || n <= 0) return null;
  return n;
}

export function parseCommitSha(sha: string | null | undefined): string | null {
  if (!sha || typeof sha !== "string") return null;
  const trimmed = sha.trim();
  if (!/^[0-9a-f]{7,40}$/i.test(trimmed)) return null;
  return trimmed;
}

function githubToken(): string | null {
  return (
    process.env.GITHUB_TOKEN?.trim() ||
    process.env.GH_MERGE_TOKEN?.trim() ||
    process.env.GH_STATUS_TOKEN?.trim() ||
    null
  );
}

/**
 * Build a GitHub merge URL from validated components.
 * GITHUB_REPOSITORY is required in production/Vercel; otherwise use the parsed repo.
 */
export function buildPullMergeUrl(
  repo: string | null | undefined,
  prNumber: unknown
): { url: string; repo: string; pr: number } | { error: string } {
  const parsed = parseGithubRepo(repo);
  if (!parsed) {
    return { error: "Review repo must look like owner/name." };
  }

  const allowed = process.env.GITHUB_REPOSITORY?.trim();
  const prodLike =
    process.env.NODE_ENV === "production" || process.env.VERCEL === "1";
  if (prodLike && !allowed) {
    return {
      error:
        "GITHUB_REPOSITORY must be set in production to pin merge targets.",
    };
  }
  if (allowed) {
    const allowedParsed = parseGithubRepo(allowed);
    if (!allowedParsed || allowedParsed.full !== parsed.full) {
      return {
        error: `Review repo ${parsed.full} is not allowed (expected ${allowed}).`,
      };
    }
  }

  const pr = parsePositiveInt(prNumber);
  if (pr === null) {
    return { error: "Review has no valid PR number." };
  }

  return {
    url: `https://api.github.com/repos/${parsed.owner}/${parsed.name}/pulls/${pr}/merge`,
    repo: parsed.full,
    pr,
  };
}

/**
 * Post/replace the Governance Quiz commit status for branch protection.
 * Soft-fails when token/repo/sha are missing — quiz UI still works.
 */
export async function setGovernanceQuizStatus(input: {
  repo: string | null | undefined;
  commitSha: string | null | undefined;
  state: QuizStatusState;
  description: string;
  targetUrl?: string | null;
}): Promise<{ ok: true } | { ok: false; error: string }> {
  const token = githubToken();
  if (!token) {
    return {
      ok: false,
      error:
        "No GITHUB_TOKEN / GH_MERGE_TOKEN / GH_STATUS_TOKEN — cannot update Governance Quiz check.",
    };
  }

  const parsed = parseGithubRepo(input.repo);
  if (!parsed) {
    return { ok: false, error: "Review repo must look like owner/name." };
  }

  const allowed = process.env.GITHUB_REPOSITORY?.trim();
  const prodLike =
    process.env.NODE_ENV === "production" || process.env.VERCEL === "1";
  if (prodLike && !allowed) {
    return {
      ok: false,
      error:
        "GITHUB_REPOSITORY must be set in production to pin status targets.",
    };
  }
  if (allowed) {
    const allowedParsed = parseGithubRepo(allowed);
    if (!allowedParsed || allowedParsed.full !== parsed.full) {
      return {
        ok: false,
        error: `Review repo ${parsed.full} is not allowed (expected ${allowed}).`,
      };
    }
  }

  const sha = parseCommitSha(input.commitSha);
  if (!sha) {
    return { ok: false, error: "Review has no valid commit SHA." };
  }

  const url = `https://api.github.com/repos/${parsed.owner}/${parsed.name}/statuses/${sha}`;
  try {
    const resp = await fetch(url, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${token}`,
        Accept: "application/vnd.github+json",
        "Content-Type": "application/json",
        "X-GitHub-Api-Version": "2022-11-28",
      },
      body: JSON.stringify({
        state: input.state,
        context: QUIZ_STATUS_CONTEXT,
        description: input.description.slice(0, 140),
        ...(input.targetUrl
          ? { target_url: input.targetUrl.slice(0, 1024) }
          : {}),
      }),
    });
    if (!resp.ok) {
      const details = await resp.text().catch(() => "");
      return {
        ok: false,
        error: `GitHub status API ${resp.status}: ${details.slice(0, 200)}`,
      };
    }
    return { ok: true };
  } catch (err) {
    return {
      ok: false,
      error: err instanceof Error ? err.message : "status update failed",
    };
  }
}

export function dashboardReviewUrl(reviewId: string): string | null {
  const base =
    process.env.GOVERNANCE_DASHBOARD_PUBLIC_URL?.trim() ||
    process.env.VERCEL_PROJECT_PRODUCTION_URL?.trim() ||
    process.env.VERCEL_URL?.trim() ||
    null;
  if (!base) return null;
  const host = base.startsWith("http") ? base : `https://${base}`;
  return `${host.replace(/\/$/, "")}/?id=${encodeURIComponent(reviewId)}`;
}
